test: Fix fragment tracking test under KUBEPROXY=1

pchaigno opened the pull request:
First commit simplifies fragment tracking test to match on 3-tuple instead of 4-tuple. The second commit fixes the test to support `KUBEPROXY=1`. See commit messages for details.

Fixes: #10929
/cc @qmonnet

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -450,11 +450,13 @@ var _ = Describe("K8sServicesTest", func() {
 				blockSize  = 5120
 				blockCount = 1
 				srcPort    = 12345
+				hostIP     = "10.0.2.15"
```
Could we get this value from elsewhere? Not sure if `node.GetExternalIPv4()` is relevant?

pchaigno replied:
We can't use that method in tests, can we?

I don't like this either. I looked at using e.g., `GetNodeIPByLabel(..., true)`, but it wasn't returning anything (the `ExternalIP` field didn't exist in kubectl's JSON output). This seems expected since other tests are also hardcoding the IP:
```
$ git grep 10.0.2.15
k8sT/Services.go:                               hostIP     = "10.0.2.15"
runtime/Policies.go:            hostIP := "10.0.2.15"
runtime/Policies.go:                    endpointID, _ := createEndpoint("ping", "10.0.2.15")
runtime/Policies.go:                            endpointID, _ := createEndpoint("ping", "10.0.2.15")
runtime/Policies.go:                    hostIP := "10.0.2.15"
```
At the very least, we could move this value to some shared file, like `helpers.K8s1IP`.

qmonnet replied:
I'm good with moving it to a helper, let's not spend too much on this. We can improve in the future if necessary, or maybe someone else will review and have an idea on how to get it in a clean way.

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -479,29 +482,55 @@ var _ = Describe("K8sServicesTest", func() {
+					newSrcIP2, err = kubectl.GetCiliumHostIPv4(context.TODO(), helpers.K8s1)
+					Expect(err).Should(BeNil(), "Cannot retrieve local cilium_host ipv4")
+				}
+			}
```
Maybe turn that block into a `if { } else { }` and do all the setting for `endpoingK8s*` and `patternInK8s*` in it? introducing the `newSrcIP*` on the kube-free path does not help readability I think.

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -479,29 +482,55 @@ var _ = Describe("K8sServicesTest", func() {
+				dstIPv4 := strings.Replace(dstIP, "::ffff:", "", 1)
+				endpointK8s1 = fmt.Sprintf("%s:%d", dstIPv4, dstPort)
+				endpointK8s2 = endpointK8s1
+			}
```
Move this to the previous `if kubeProxy`?

qmonnet reviewed (commented):
Thanks a lot! This looks great and looks correct in regard to my understanding of the issue.

The test does become more complicated to read, but I see no ideal solution. We could decouple the test function for the kubeproxy/kube-free cases, but we would end up with something longer in total and with a lot of duplicated code. Still, a few suggestions inline.

Could you please explain what you mean about the `fromNode == helpers.K8s2` case? I didn't get your point :/.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -479,29 +482,55 @@ var _ = Describe("K8sServicesTest", func() {
+					newSrcIP2 = hostIP
+				} else {
+					newSrcIP1 = hostIP
+					newSrcIP2, err = kubectl.GetCiliumHostIPv4(context.TODO(), helpers.K8s1)
```
For the `fromNode == helpers.K8s2` case:
Here we are basically inverting `newSrcIP1` and `newSrcIP2` based on whether the destination is on `k8s1` or `k8s2`. I'm expecting this does not actually depend on `k8s1` vs. `k8s2`, but on whether the destination is on the same node as the source or not. So this depends on `fromNode`.

However, as far as I can see, the tests do not make use of the `fromNode` parameter, so I'm wondering if we should just remove it and consider we always send from `k8s1`.

qmonnet replied:
Ok, thanks. Yeah, you're probably right, we can try that.

pchaigno committed:
test: Simplify fragment tracking test

The fragment tracking test sends a couple fragments to a K8s service and
checks the ctmap content to ensure they were properly sent and received.
It used to filter based on the 4-tuple. That is however making the test
convoluted when adding support for kube-proxy to the test (new commit).

This commit therefore switches to filtering based on the 3-tuple (source
port, destination IP and port). Support for sending from k8s2 is also
removed since it wasn't used.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Fix fragment tracking test under KUBEPROXY=1

The fragment tracking test currently fails when kube-proxy is enabled
because the destination IP address and port are sometimes not DNATed.
The awk filter on the ctmap dump output fails in that case.

This commit fixes this issue by checking if kube-proxy is enabled and
adapting the destination on which to match as a result.

Fixes: #10929
Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet reviewed (approved):
Looks great, thanks a lot for debugging and patching this!
